File: forms/advance-decision-to-refuse-treatment/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/validity_grader.rs
use super::types::{AssessmentData, FiredRule, ValidityStatus};
use super::validity_rules::all_rules;

/// Pure function: evaluates all validity rules against ADRT data.
///
/// Determines validity status based on which rules fire:
/// - If ANY critical or required rule fires: check draft heuristic → draft or invalid
/// - If ONLY recommended rules fire: complete
/// - If NO rules fire: valid
pub fn calculate_validity(data: &AssessmentData) -> (ValidityStatus, Vec<FiredRule>) {
    let mut fired_rules = Vec::new();

    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                severity: rule.severity.to_string(),
            });
        }
    }

    let has_critical = fired_rules.iter().any(|r| r.severity == "critical");
    let has_required = fired_rules.iter().any(|r| r.severity == "required");
    let has_recommended = fired_rules.iter().any(|r| r.severity == "recommended");

    let validity_status = if has_critical || has_required {
        if is_likely_draft(data) {
            "draft".to_string()
        } else {
            "invalid".to_string()
        }
    } else if has_recommended {
        "complete".to_string()
    } else {
        "valid".to_string()
    };

    (validity_status, fired_rules)
}

/// Heuristic: if core identification fields are empty, treat as draft rather than invalid.
fn is_likely_draft(data: &AssessmentData) -> bool {
    data.personal_information.full_legal_name.trim().is_empty()
        && data.personal_information.date_of_birth.is_empty()
        && data.legal_signatures.patient_signature != "yes"
}

File: forms/advance-decision-to-refuse-treatment/front-end-form-with-svelte/src/lib/engine/validity-grader.ts
import type { AssessmentData, ValidityStatus, FiredRule } from './types';
import { validityRules } from './validity-rules';

/**
 * Pure function: evaluates all validity rules against ADRT data.
 *
 * Determines validity status based on which rules fire:
 * - If ANY critical rule fires (life-sustaining treatment requirements not met): Invalid
 * - If ANY required rule fires (missing legal requirements): Invalid
 * - If ONLY recommended rules fire: Complete (all required sections filled but could be improved)
 * - If NO rules fire: Valid (fully legally compliant)
 *
 * A "fired rule" means the rule's condition evaluated to true, indicating
 * that something is MISSING or WRONG (i.e., the rule detects a deficiency).
 */
export function calculateValidity(data: AssessmentData): {
	validityStatus: ValidityStatus;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	for (const rule of validityRules) {
		try {
			if (rule.evaluate(data)) {
				firedRules.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					severity: rule.severity
				});
			}
		} catch (e) {
			// Rule evaluation failed - log for debugging but continue
			console.warn(`Validity rule ${rule.id} evaluation failed:`, e);
		}
	}

	const hasCritical = firedRules.some((r) => r.severity === 'critical');
	const hasRequired = firedRules.some((r) => r.severity === 'required');
	const hasRecommended = firedRules.some((r) => r.severity === 'recommended');

	let validityStatus: ValidityStatus;
	if (hasCritical || hasRequired) {
		// Check if most fields are empty (draft vs invalid)
		const isDraft = isLikelyDraft(data);
		validityStatus = isDraft ? 'draft' : 'invalid';
	} else if (hasRecommended) {
		validityStatus = 'complete';
	} else {
		validityStatus = 'valid';
	}

	return { validityStatus, firedRules };
}

/** Heuristic: if core identification fields are empty, treat as draft rather than invalid. */
function isLikelyDraft(data: AssessmentData): boolean {
	return (
		data.personalInformation.fullLegalName.trim() === '' &&
		data.personalInformation.dateOfBirth === '' &&
		data.legalSignatures.patientSignature !== 'yes'
	);
}
